src/main/java/com/mongodb/hibernate/translate/attachment/Attachment.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright 2024-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.hibernate.translate.attachment;
+
+import static java.lang.String.format;
+
+import com.mongodb.hibernate.internal.VisibleForTesting;
+import org.jspecify.annotations.Nullable;
+
+/**
+ * Manages global state while translating SQL AST via {@link org.hibernate.sql.ast.SqlAstWalker}.
+ *
+ * <p>{@code SqlAstWalker} is based on {@code visitor} pattern without returning value from the various visitor methods.
+ * The original goal of {@code SqlAstWalker} is for concatenating SQL string internally; but when translating SQL AST
+ * into MQL, it is highly desirable to return value (e.g. {@link com.mongodb.hibernate.translate.mongoast.AstNode}). To
+ * avoid such dilemma, some attachment/detachment mechanism is implemented as follows:
+ *
+ * <ul>
+ *   <li>when some visitor method needs to return some value, it attaches the value by invoking
+ *       {@link #attach(AttachmentKey, Object)}.
+ *   <li>when some coordinating detacher needs to access the value attached above, it invokes
+ *       {@link #expect(AttachmentKey, Runnable)}; internally the {@code Runnable} will attach value by the above
+ *       attacher.
+ *   <li>both attacher and detacher should provide identical {@link AttachmentKey} to ensure data transferring safety.
+ * </ul>
+ *
+ * @see org.hibernate.sql.ast.SqlAstWalker
+ * @see AttachmentKey
+ */
+@SuppressWarnings("unchecked")
+public final class Attachment {
+
+    private @Nullable AttachmentKey<?> attachmentKey;
+    private @Nullable Object attachedValue;
+
+    @Override
+    public String toString() {
+        return "Attachment{" + "attachmentKey=" + attachmentKey + ", attachedValue=" + attachedValue + "}";
+    }
+
+    @VisibleForTesting(otherwise = VisibleForTesting.AccessModifier.PRIVATE)
+    boolean isBlank() {
```

Comment:
What is the purpose of this method, even for tests?

Reply:
A key feature of Attachment is to support embedded attaching (there is a unit testing case to demonstrate that). I used this method to verify that after usage, its initial blank state should be reverted back. That covers the case that some dangling data lingers, e.g.:

- `Runnalbe` throws an exception and our `Attachment` didn't restore previous state;
- deeply nested attachting didn't work as expected due to defective "previous state restoration"

tbh, all these Attachment complexity is really difficult to understand and use. Without a perfect unit testing coverage, it is so error-prone.

Reply:
> tbh, all these Attachment complexity is really difficult to understand and use. 

Is there a better alternative?

Reply:
we have an obvious choice here. A common issue with common solution. E.g. see https://stackoverflow.com/questions/1711797/how-to-store-state-on-a-node-for-a-visitor-pattern

It is also explained in detail in one book I read about ANTLR library (https://pragprog.com/titles/tpantlr2/the-definitive-antlr-4-reference/). Ultimately Hibernate uses such visitor pattern is simply due to ANTLR library (which generates visitor API as the first vistior API in Hibernate; Hibernate simulates it to accommodate its SQL translation step. Actually Hibernate uses such visitor walker more than once, including:

1. SQM AST -> SQL AST (e.g. https://github.com/hibernate/hibernate-orm/blob/main/hibernate-core/src/main/java/org/hibernate/query/sqm/sql/BaseSqmToSqlAstConverter.java). Return type is used for the various visitor methods for Hibernate has internal SQL AST model. Below is the method to convert `SqmUpdateStatement` to `SqmUpdateStatement`:
<img width="735" alt="Screenshot 2024-12-09 at 2 36 33 PM" src="https://github.com/user-attachments/assets/562bfbac-3554-4cc0-b2c4-5c09f93eb0ce">


2. SQL AST -> SQL: given the target is flat SQL string, so it is natural Hibernate opts for void visitor methods, which is not enough for us for we have intermediate mongo AST

I explored whether we could make the above second item to return type as in SQM AST -> SQL AST case, but it has long-term maintainability issue so we'd better adhere to Hibernate's existing walker API without customization.

The most natural way to solve the issue is as follows:

- maintain a global map from SQL AST tree node reference to its corresponding visitor method's returned value; ANTLR book even provides the detail to use `IdentityHashMap` instead of `HashMap` for opt.
- when we visit an AST node in its visit method, we adopt post-order traversal by visiting its children nodes first, then we look up in the global map to get their values, then we finish computing the returned value for the current node and store it into global map as well.

There is still complexity like ensuring type safety but I want to say it enjoys the biggest benefit: easy understanding and easy maintaining. Suppose I am community contributor, I would feel easy to get the hang of the subtlety without firstly scratching his/her haris to figure out the complicated `Attachment` mechanism (including its subtle old state storation detail, which relies on stacktrace local variable to play hand in hand with recursion or depth-first tree travseral).